### PR TITLE
Add interesting example to welcome page

### PR DESCRIPTION
**Pull request by JackStouffer**
A new attempt at showing off cool D features from moment one.

I think we do a great job of pulling in new-er programmers with Ali's book and this tour. But experienced programmers need a hook to sink their time into D.

### CODE REVIEW

**Comment by JackStouffer** [comment]
CI failure likely due to dub outage

**Comment by JackStouffer** [comment]
BTW, this is one of the examples that blew people's minds at CppCon when Andrei gave a talk there on ranges in 2009.

**Comment by wilzbach** [comment]
> CI failure likely due to dub outage

Yeah :/
Put you on the CC to a mail I sent to Martin about preventing this the next time

> BTW, this is one of the examples that blew people's minds at CppCon when Andrei gave a talk there on ranges in 2009.

Okay first of all: no matter whether you are successful in pushing this or not, I really like the example and initiative and the tour could use more of these sweet snippets!
So in case of failure, I suggest using it for this page: https://tour.dlang.org/tour/en/basics/arrays

About having it on the front page: in short I think it's a bit too messy:
- the `writeln` calls feel a bit redundant, I would prefer sth. that uses more of the range API, e.g. `arrs.each!writeln`
- it's rather "hard" to see (visually proof) and understand why no allocation is happening here without knowing about slices and ranges in D

However, I do see your point of showing off and providing a sneak peek in the beginning.
Maybe we can use sth. like the snippet on dlang.org that uses UCFS and ranges to the full extent?

For voting and as @stonemaster is currently busy, we might want to consider [his vote](https://github.com/stonemaster/dlang-tour/pull/241#issuecomment-225423604).

**Comment by wilzbach** [comment]
> I suggest using it for this page: https://tour.dlang.org/tour/en/basics/arrays

edit: or this http://tour.dlang.io/tour/en/basics/slices

File: welcome/welcome-to-d.md [hunk 1]
```diff
@@ -48,10 +48,21 @@ and we welcome pull requests making this tour even better.
...
+    
+    // More advanced example for expirenced programmers
+    // Take three arrays, and without allocating any new memory
+    // sort across all of them inplace
+    int[] arr1 = [4, 9, 7];
+    int[] arr2 = [5, 2, 1];
+    int[] arr3 = [6, 8, 3];
+    sort(chain(arr1, arr2, arr3));
+    writeln(arr1); writeln(arr2); writeln(arr3);
 }
```
Comment (ZombineDev) [inline welcome/welcome-to-d.md]: Use `writefln("%s\n%s\n%s", arr1, arr2, arr3)` or `writeln(arr1, arr2, arr3)`

File: welcome/welcome-to-d.md [hunk 2]
```diff
@@ -48,10 +48,21 @@ and we welcome pull requests making this tour even better.
...
+import std.range;
 
-// Let's get going!
 void main()
 {
+    // Let's get going!
     writeln("Hello World!");
+    
+    // A cool example for expirenced programmers:
+    // Take three arrays, and without allocating any new memory,
```
Comment (wilzbach) [inline welcome/welcome-to-d.md]: I think this leads to

> object.Exception@source/contentprovider.d(288): Source code line length exceeds 48 limit in 'Welcome to D':     // A cool example for expirenced programmers:

File: welcome/welcome-to-d.md [hunk 3]
```diff
@@ -48,10 +48,21 @@ and we welcome pull requests making this tour even better.
...
+import std.algorithm;
+import std.range;
 
-// Let's get going!
 void main()
 {
+    // Let's get going!
     writeln("Hello World!");
+    
+    // A cool example for expirenced programmers:
```
Comment (wilzbach) [inline welcome/welcome-to-d.md]: I wouldn't use the word "cool", sounds a bit like self-praise ...

**Comment by JackStouffer** [comment]
Well, the idea is not to present the reader with all of the answers, but to give them a hook which will cause them to ask questions and become interested.

I made this PR after I went through the Rust tutorial. They definitely have a leg up on us in the area of making new users interested from the start. They do this by advertising what their compile time safety checks can accomplish, and leave the implementation details for the reader to figure out later. They get users interested by telling them about the things that Rust can do that no one else can.

Similarly, I would like to see someone attempt this in C++ with iterators or Python and JS with their generator syntax.

I'll add a link to the range algorithms page at the bottom of the code.

**Comment by ZombineDev** [comment]
Alright, I think we have reached consensus that this strikes the right balance for now. We can always improve things later. Thanks @JackStouffer for your contribution!

**Pull request merged**